debug/debugger.go

```diff
@@ -0,0 +1,859 @@
+		}
+
+		tracer.resultSet = rs
+		s.result(t, rs)
+	}()
+
+	if err := s.start(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func readInput(path string) (interface{}, error) {
+	path, err := fileurl.Clean(path)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var input interface{}
+	if err := json.Unmarshal(data, &input); err != nil {
+		return nil, err
+	}
+
+	return input, nil
+}
+
+type session struct {
+	d              *debugger
+	properties     LaunchProperties
+	threads        []*thread
+	frames         []*stackFrame
+	framesByThread map[ThreadID][]*stackFrame
+	breakpoints    *breakpointCollection
+	ctx            context.Context
+	cancel         context.CancelFunc
+	varManager     *variableManager
+}
+
+func newSession(ctx context.Context, debugger *debugger, varManager *variableManager, props LaunchProperties, threads []*thread) *session {
+	ctx, cancel := context.WithCancel(ctx)
+	s := &session{
+		d:              debugger,
+		varManager:     varManager,
+		properties:     props,
+		threads:        threads,
+		frames:         []*stackFrame{},
+		framesByThread: map[ThreadID][]*stackFrame{},
+		breakpoints:    newBreakpointCollection(),
+		ctx:            ctx,
+		cancel:         cancel,
+	}
+
+	for _, t := range threads {
+		t.eventHandler = s.handleEvent
+	}
+
+	return s
+}
+
+func (s *session) start() error {
+	if s == nil {
+		return fmt.Errorf("no active debug session")
+	}
+
+	for _, t := range s.threads {
+		t := t
+		go func() {
+			s.d.logger.Debug("Thread %d started", t.id)
+			s.d.sendEvent(Event{Type: ThreadEventType, Thread: t.id, Message: "started"})
+			if err := t.run(s.ctx); err != nil {
+				s.d.logger.Error("Thread %d failed: %v", t.id, err)
+			}
+			s.d.logger.Debug("Thread %d stopped", t.id)
+			s.d.sendEvent(Event{Type: ThreadEventType, Thread: t.id, Message: "exited"})
+
+			allStopped := true
+			for _, t := range s.threads {
+				if !t.done() {
+					allStopped = false
+					break
+				}
+			}
+
+			if allStopped {
+				s.d.logger.Debug("All threads stopped")
+				s.d.sendEvent(Event{Type: TerminatedEventType})
+			}
+		}()
+	}
+
+	return nil
+}
+
+func (s *session) thread(id ThreadID) (*thread, error) {
+	if s == nil {
+		return nil, fmt.Errorf("no active debug session")
+	}
+
+	index := int(id - 1)
+	if index < 0 || index >= len(s.threads) {
+		return nil, fmt.Errorf("invalid thread id: %d", id)
+	}
+	return s.threads[index], nil
+}
+
+func (s *session) Resume(threadID ThreadID) error {
+	if s == nil {
+		return fmt.Errorf("no active debug session")
+	}
+
+	t, err := s.thread(threadID)
+	if err != nil {
+		return err
+	}
+
+	return t.resume()
+}
+
+func (s *session) ResumeAll() error {
+	if s == nil {
+		return fmt.Errorf("no active debug session")
+	}
+
+	for _, t := range s.threads {
+		if err := t.resume(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *session) StepOver(threadID ThreadID) error {
```

Comment:
Nit: `StepOver`, `StepOut` and `StepIn` have some duplicate code which we could move in a helper.